core/consensus/grandpa/impl/launcher_impl.cpp

```diff
@@ -177,20 +180,50 @@ namespace kagome::consensus::grandpa {
   void LauncherImpl::onVoteMessage(const VoteMessage &msg) {
     auto current_round = current_round_;
     auto current_round_number = current_round->roundNumber();
-    // TODO(xDimon): Implement mechanism to check if signer is known
-    if (msg.round_number == current_round_number) {
-      visit_in_place(
-          msg.vote.message,
-          [&current_round, &msg](const PrimaryPropose &primary_propose) {
-            current_round->onPrimaryPropose(msg.vote);
-          },
-          [&current_round, &msg](const Prevote &prevote) {
-            current_round->onPrevote(msg.vote);
-          },
-          [&current_round, &msg](const Precommit &precommit) {
-            current_round->onPrecommit(msg.vote);
-          });
+
+    // ensure we are in current round
+    if (msg.round_number != current_round_number) {
+      return;
     }
+
+    // get block info
+    auto blockInfo = visit_in_place(msg.vote.message, [](const auto &vote) {
+      return BlockInfo(vote.block_number, vote.block_hash);
+    });
+
+    // get authorities
+    const auto &weighted_authorities_res =
+        grandpa_api_->authorities(primitives::BlockId(blockInfo.block_number));
```

Comment:
`authorities` method does not reset state. Hence, it should be updated to reset to the state corresponding to the block defined by its id

Reply:
Really? In [grandpa_api](https://github.com/soramitsu/kagome/blob/e43fd81a1c0924787502b0f4017aa27b91a3af23/core/runtime/binaryen/runtime_api/grandpa_impl.cpp#L39) this method executing on ephemeral batch. Or I catched it wrong?

Reply:
indeed, you are right. Sorry, cannot get used to that ephemeral batches approach